mdf_matio/validator.py

```diff
@@ -0,0 +1,397 @@
+                if char.isupper():  # Start of new element symbol
+                    str_of_elem += " " + char
+                elif char.islower():  # Continuation of symbol
+                    str_of_elem += char
+                # Anything else is not an element (numbers, whitespace, etc.)
+
+            # Split elements in string (on whitespace), make unique and JSON-serializable
+            list_of_elem = list(set(str_of_elem.split()))
+            # Ensure deterministic results
+            list_of_elem.sort()
+            # Currently deprecated
+            # If any "element" isn't in the periodic table,
+            # the composition is likely not a chemical formula and should not be parsed
+#                if all([elem in DICT_OF_ALL_ELEMENTS.values() for elem in list_of_elem]):
+#                    record["elements"] = list_of_elem
+
+            rc_md["material"]["elements"] = list_of_elem
+        elif rc_md["material"].get("elemental_proportions"):
+            rc_md["material"]["elements"] = [rc_md["material"]["elemental_proportions"].keys()]
+            rc_md["material"]["elements"].sort()
+
+        # BLOCK: custom
+        # Make all values into strings
+        if rc_md.get("custom"):
+            new_custom = {}
+            for key, val in rc_md["custom"].items():
+                new_custom[key] = str(val)
+            rc_md["custom"] = new_custom
+
+        # Require strict JSON
+        try:
+            json.dumps(rc_md, allow_nan=False)
+        except (ValueError, json.JSONDecodeError) as e:
+            return {
+                "success": False,
+                "error": "Invalid record JSON: {}".format(str(e)),
+                "details": repr(e)
+                }
+
+        # Remove null/None values
+        rc_md = _remove_nulls(rc_md, self.__allowed_nulls)
+
+        # Validate against schema
+        try:
+            jsonschema.validate(rc_md, schema, resolver=resolver)
+        except jsonschema.ValidationError as e:
+            return {
+                "success": False,
+                "error": "Invalid record metadata: " + str(e).split("\n")[0],
+                "details": str(e)
+                }
+
+        # Write out to file
+        json.dump(rc_md, self.__tempfile)
+        self.__tempfile.write("\n")
+
+        # Return results
+        return {
+            "success": True
+            }
+
+    def get_finished_dataset(self):
+        """Retrieve finished dataset, in a generator."""
+        if self.__dataset is None:
+            raise ValueError("Dataset not started")
+        elif self.__finished:
+            raise ValueError("Dataset already finished")
+
+        self.__indexed_files = []
+        self.__finished = True
+
+        self.__tempfile.seek(0)
+        yield self.__dataset
+        for line in self.__tempfile:
+            yield json.loads(line)
+
+        self.__tempfile.close()
+        self.__dataset = None
+        return
+
+    def status(self):
```

Comment:
If you agree this class need not hold on to the records, then we can kill this function and `get_finished_dataset`